pymc_experimental/inference/jax_find_map.py

```diff
@@ -0,0 +1,450 @@
+import logging
+
+from collections.abc import Callable
+from typing import Literal, cast
+
+import arviz as az
+import jax
+import numpy as np
+import pymc as pm
+import pytensor
+import pytensor.tensor as pt
+import xarray as xr
+
+from arviz import dict_to_dataset
+from better_optimize import minimize
+from better_optimize.constants import minimize_method
+from pymc.backends.arviz import (
+    coords_and_dims_for_inferencedata,
+    find_constants,
+    find_observations,
+)
+from pymc.blocking import DictToArrayBijection, RaveledVars
+from pymc.initial_point import make_initial_point_fn
+from pymc.model.transform.conditioning import remove_value_transforms
+from pymc.model.transform.optimization import freeze_dims_and_data
+from pymc.sampling.jax import get_jaxified_graph
+from pymc.util import get_default_varnames
+from pytensor.tensor import TensorVariable
+from scipy import stats
+from scipy.optimize import OptimizeResult
+
+_log = logging.getLogger(__name__)
+
+
+def get_near_psd(A: np.ndarray) -> np.ndarray:
+    """
+    Compute the nearest positive semi-definite matrix to a given matrix.
+
+    This function takes a square matrix and returns the nearest positive
+    semi-definite matrix using eigenvalue decomposition. It ensures all
+    eigenvalues are non-negative. The "nearest" matrix is defined in terms
+    of the Frobenius norm.
+
+    Parameters
+    ----------
+    A : np.ndarray
+        Input square matrix.
+
+    Returns
+    -------
+    np.ndarray
+        The nearest positive semi-definite matrix to the input matrix.
+    """
+    C = (A + A.T) / 2
+    eigval, eigvec = np.linalg.eig(C)
+    eigval[eigval < 0] = 0
+
+    return eigvec @ np.diag(eigval) @ eigvec.T
+
+
+def _get_unravel_rv_info(optimized_point, variables, model):
+    cursor = 0
+    slices = {}
+    out_shapes = {}
+
+    for i, var in enumerate(variables):
+        raveled_shape = np.prod(optimized_point[var.name].shape).astype(int)
+        rv = model.values_to_rvs.get(var, var)
+
+        idx = slice(cursor, cursor + raveled_shape)
+        slices[rv] = idx
+        out_shapes[rv] = tuple(
+            [len(model.coords[dim]) for dim in model.named_vars_to_dims.get(rv.name, [])]
+        )
```

Comment:
What if an rv doesn't have dims specified?

Reply:
is there already a helper in pymc to figure out shapes of an rv?

Reply:
If you froze the dims you should have everything in `rv.type.shape`. Otherwise we usually use `model.initial_point`